### PR TITLE
Kubectl ko diagnose perf

### PR DESCRIPTION
**Pull request by changluyi**
- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features
- Bug fixes
- Docs
- Tests

1. kubectl ko add perf
2. add TCP/UDP connectivity check
3.  add kubectl ko diagnose subnet

### Which issue(s) this PR fixes:
Fixes #(issue-number)

### <samp>🤖 Generated by Copilot at a05e06f</samp>

This pull request adds TCP and UDP connectivity checks to the kube-ovn network. It modifies the `daemon`, `pinger`, and `util` packages to implement and configure the checks, and updates the `test-server.yaml` file to support `qperf` testing. It also adds TCP and UDP listeners to the `cniserver.go` and `pinger.go` files to enable the checks between nodes and pods.

### <samp>🤖 Generated by Copilot at a05e06f</samp>

> _Oh we are the `kube-ovn` crew and we work on the net_
> _We listen for TCP and UDP on the ports that we set_
> _We use the `util` package and the `config` flags_
> _And we heave away, me hearties, heave away on the lags_

### <samp>🤖 Generated by Copilot at a05e06f</samp>

*  Add TCP and UDP connectivity checks between nodes and pods in the kube-ovn network ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-2eecf9a49ec68a8fcdcaa8a202d707c2a72418faf0dbb84edf615440c234adbbR112-R126), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-cc7c97c269c4de582c3c4aaaa10bf1dfc8507245babb89bb7cbc2877f6f61a33R39-R52), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-ef5e19d2a06e783042497957187c9a7df962e6db67629dfbe8075e6f12c02704R95-R108), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-ef5e19d2a06e783042497957187c9a7df962e6db67629dfbe8075e6f12c02704R160-R173), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-a1ffb52e88506b4737c752182fb157473a2575e8e7bb356ed709d2316141a3cdR529-R612))
  * Add two fields to store the port numbers for the TCP and UDP checks in the `Configuration` type in `config.go` files of the `daemon` and `pinger` packages ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-fde9321dc0cca23d5eacc02fbb93b4c7979afdb901f7bd0ce31af16fe5ba1073R63-R64), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-56f139df697aad97197df93aedd368692be48a0ef3d6a4399b7f17fea0c6c8fcL53-R63))
  * Add two command-line flags to set the port numbers for the TCP and UDP checks in the `ParseFlags` functions in `config.go` files of the `daemon` and `pinger` packages ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-fde9321dc0cca23d5eacc02fbb93b4c7979afdb901f7bd0ce31af16fe5ba1073R99-R100), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-fde9321dc0cca23d5eacc02fbb93b4c7979afdb901f7bd0ce31af16fe5ba1073R152-R153), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-56f139df697aad97197df93aedd368692be48a0ef3d6a4399b7f17fea0c6c8fcR127-R128))
  * Add two goroutines to start TCP and UDP listeners on the specified address and port in the `main` functions of the `cniserver.go` and `pinger.go` files of the `daemon` and `pinger` packages, using the `TCPConnectivityListen` and `UDPConnectivityListen` functions from the `util` package ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-2eecf9a49ec68a8fcdcaa8a202d707c2a72418faf0dbb84edf615440c234adbbR112-R126), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-cc7c97c269c4de582c3c4aaaa10bf1dfc8507245babb89bb7cbc2877f6f61a33R39-R52))
  * Add two calls to check the TCP and UDP connectivity to the node and pod IP address and port in the `pingNodes` and `pingPods` functions of the `ping.go` file of the `pinger` package, using the `TCPConnectivityCheck` and `UDPConnectivityCheck` functions from the `util` package, and log the results ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-ef5e19d2a06e783042497957187c9a7df962e6db67629dfbe8075e6f12c02704R95-R108), [link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-ef5e19d2a06e783042497957187c9a7df962e6db67629dfbe8075e6f12c02704R160-R173))
  * Add four functions to implement the logic for checking and listening for TCP and UDP connectivity, using the `net` package, in the `net.go` file of the `util` package ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-a1ffb52e88506b4737c752182fb157473a2575e8e7bb356ed709d2316141a3cdR529-R612))
  * Add the `time` package to the import list in the `net.go` file of the `util` package, to use it for setting timeouts and deadlines for the TCP and UDP checks ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-a1ffb52e88506b4737c752182fb157473a2575e8e7bb356ed709d2316141a3cdR12))
* Add a test server pod on the control plane node, which can be used for testing the TCP and UDP connectivity checks from other nodes and pods, in the `test-server.yaml` file in the `test` folder ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-aaa53b027bdb6b3327d2a742e325d6ab2ce9590b2363be945280c1667b22e99bR25-R46))
  * Modify the `command` and `args` fields of the `test-server` container to run `qperf` in the background, before running the `test-server.sh` script, to enable `qperf` testing for the TCP and UDP checks ([link](https://github.com/kubeovn/kube-ovn/pull/2915/files?diff=unified&w=0#diff-aaa53b027bdb6b3327d2a742e325d6ab2ce9590b2363be945280c1667b22e99bR12-R16))

### CODE REVIEW

File: dist/images/kubectl-ko [hunk 1]
```diff
@@ -1082,6 +1089,171 @@ log(){
   echo "Collected files have been saved in the directory $PWD/kubectl-ko-log "
 }
 
+perfTimes=5
+
+perf(){
+  curl -O https://raw.githubusercontent.com/kubeovn/kube-ovn/master/test/server/test-server.yaml
```
Comment (oilbeater) [inline dist/images/kubectl-ko:1095]: 需要默认考虑离线场景，集群大概率不能连外网

File: cmd/daemon/cniserver.go [hunk 2]
```diff
@@ -109,6 +109,21 @@ func CmdMain() {
 			}
 		}
 	}
+
```
Comment (oilbeater) [inline cmd/daemon/cniserver.go:112]: 需要配置能否关闭，不然会存在host网络端口安全检查的问题

File: test/server/test-server.yaml [hunk 3]
```diff
@@ -9,6 +9,33 @@ spec:
     - name: test-server
       image: kubeovn/test:v1.12.0
       imagePullPolicy: IfNotPresent
+      command: ["sh", "-c"]
+      args:
+        - |
+          qperf &
+          ./test-server.sh
+  nodeSelector:
```
Comment (oilbeater) [inline test/server/test-server.yaml:17]: test-server.yaml 之前是为 kind 环境特殊配置的 nodeselector 需要考虑更通用的场景

File: dist/images/kubectl-ko [hunk 4]
```diff
@@ -1082,6 +1089,171 @@ log(){
...
+
+  nodes=($(kubectl get node --no-headers -o custom-columns=NAME:.metadata.name))
+  if [[ ${#nodes} = 1 ]]; then
```
Comment (oilbeater) [inline dist/images/kubectl-ko:1098]:
```suggestion
  if [[ ${#nodes} -eq 1 ]]; then
```

File: dist/images/kubectl-ko [hunk 5]
```diff
@@ -1082,6 +1089,171 @@ log(){
...
+    sed -i "s/kube-ovn-control-plane/${nodes[0]}/g" test-server.yaml
+    sed -i "s/kube-ovn-worker/${nodes[1]}/g" test-server.yaml
+  fi
+
+  kubectl apply -f test-server.yaml
+
+  isfailed=true
+  for i in {0..59}
+  do
+    clientPod=$(kubectl get pod test-client | awk '{if($2=="1/1" && $3=="Running") print $1}')
```
Comment (oilbeater) [inline dist/images/kubectl-ko:1114]: 可以用 kubectl  wait 简化逻辑

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: test/server/test-server.yaml (hunk 3 above)
Comment (changluyi) [reply test/server/test-server.yaml:17]: 这个在脚本里面有处理 会去sed 替换nodeselector字段。

File: dist/images/kubectl-ko (hunk 1 above)
Comment (changluyi) [reply dist/images/kubectl-ko]: 已修复以下,并增加kubectl ko diagnose

File: cmd/daemon/cniserver.go (hunk 2 above)
Comment (changluyi) [reply cmd/daemon/cniserver.go:112]: fixed

File: dist/images/kubectl-ko (hunk 4 above)
Comment (changluyi) [reply dist/images/kubectl-ko]: fixed

File: dist/images/kubectl-ko (hunk 5 above)
Comment (changluyi) [reply dist/images/kubectl-ko]: fixed

**Review by oilbeater — APPROVED** [review]

**Pull request merged**